### ISSUE TITLE
Consider a new default scheduler

### ISSUE THREAD

**Issue by carmstrong** [deis/deis#4222]
Deis now has support for Docker Swarm, Apache Mesos, and Google Kubernetes as application schedulers. With the known limitations of fleet (primarily, not being a resource-aware scheduler), we should investigate using a different scheduler as our default.

**Comment by rimusz** [comment]
+1 for Kubernetes

**Comment by Blystad** [comment]
+1 for Kubernetes here as well

**Comment by krancour** [comment]
Popular opinion seems to favor kubernetes.  I'm going to link here to known (so far) issues with the k8s tech preview that probably need addressing before it can be considered as the new default scheduler.

So far:

* #4264 
* #4263

**Comment by sstarcher** [comment]
Before switching to k8s it should probably be deployed with multiple api servers.  As of now if the api server goes down it will cause everything scheduled through k8s to die.

**Comment by rimusz** [comment]
@sstarcher if api server goos down, nothing scheduled through k8s dies, pods still will be running.
But is a good idea to have HA k8s master setup

**Comment by sstarcher** [comment]
@rimusz In a single k8s master setup if the master server dies and the k8s master has to be brought back up it will reschedule all of the pods.

**Comment by krancour** [comment]
Don't worry.  I'll be testing what tolerance the system has for the death of each k8s component.  Ultimately, we will roll out something HA if we promote k8s to the default scheduler.

**Comment by rimusz** [comment]
@sstarcher I was playing with taking off ectd and master, just master, just etcd and pods stayed as it is.
a while ago somewhere  in k8s docs I found that pods will not be rescheduled if master is not available and comes back.

**Comment by krancour** [comment]
Adding #4344 to the list of considerations.

**Comment by krancour** [comment]
Updating where we currently stand with this (if k8s is the most eligible candidate as the new default scheduler).

Related issues... resolved if struck through:

* ~~#4264 / #3633~~ (fixed via #4291)
* #4263 
* #4158 (likely caused by #4269 and _not_ a blocker)
* #4344 (fix in progress)

Also now investigating:

https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/admin/high-availability.md

**Comment by sstarcher** [comment]
@krancour Something that is not really discussed in the k8s HA docs that should be discussed is ETCD continued availability over cluster churn.  I'm going to talk about this specifically in an AWS environment since that's how we are running our cluster.

When a node running ETCD is terminated it is not removed from the nodes list.  If you are using the discovery URL for ETCD you have to tell it the initial cluster size.  If you don't specify the initial cluster size it assumes it is 3.  If the initial cluster size is 3 and you launch 5 nodes that connect through the ETCD discovery URL the default behavior is to have the first 3 nodes that connect become the nodes that control the quorum and the other 2 nodes to be proxy nodes.  If a proxy node dies and AWS fires off an autoscaling event and brings up a new node it does not cause a problem, but if a master node dies it will stay a part of the cluster and the new node that comes up will become a proxy node as this is now the 6th node in the cluster.  After the node comes up you will be left with 2 master nodes and if a second master node dies you will then have a broken cluster.

TLDR;
* ETCD Discovery URL expects a static cluster size
* If a node is terminated it must be removed from the ETCD Discovery URL cluster listing or any additional nodes will be proxies
* Something must remove the terminated node prior to the autoscaling event of a new server

Some light reading on the proxy-fallback mechanism https://github.com/coreos/etcd/issues/1457
Some light reading on Adding/Removing Nodes from ETCD https://github.com/coreos/etcd/blob/master/Documentation/runtime-configuration.md#replace-a-failed-machine

**Comment by krancour** [comment]
@sstarcher that's not really a k8s-specific problem.

**Comment by bacongobbler** [comment]
I believe we [already have this documented](http://docs.deis.io/en/latest/managing_deis/add_remove_host/#removing-the-host-from-etcd). It's a little tiny footnote, but it's there. If we migrate to kubernetes we will make sure that this tidbit is still in there so administrators will know how to add and remove hosts from the cluster as we do with ceph and etcd today. Does that seem fair, or is there something else that needs more clarification?

**Comment by jgmize** [comment]
#4346 is another related issue for k8s.

**Comment by sstarcher** [comment]
@bacongobbler My mistake I originally thought the graceful-shutdown was mainly to cleanup ceph.  I did not realize it cleans up etcd.  I'll test it out later to verify it covers my use cases.

**Comment by krancour** [comment]
@jgmize #4346 is likely not scheduler-specific.

**Comment by krancour** [comment]
Tech previews of non-Fleet schedulers have been removed from the master branch, so I'm going to close this issue.  For anyone following and wishing to use k8s still, have a look at https://github.com/deis/workflow.
